### ISSUE TITLE
Restic on APFS needs small adjustments

### ISSUE THREAD

**Issue by lloeki** [restic/restic#1087]
## Output of `restic version`

```
$  ./restic version
debug enabled
restic 0.6.1 (v0.6.1-73-gaabe2a0-dirty)
compiled with go1.8.1 on darwin/amd64
```

## How did you start restic exactly? (Include the complete command line)

```
$ gb test cmds/restic
```

## What backend/server/service did you use?

Local FS.

## Expected behavior

Tests should pass.

## Actual behavior

Using 10.13 beta on APFS, there is a precision discreptancy. I don't know what the deal is with the lchown.

```
$ gb test cmds/restic
# cmds/restic
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-242244809
Don't forget to umount after quitting!
Now serving the repository at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-mount-055822315
Don't forget to umount after quitting!
error for testdata/0/0/9/37: Lstat: lstat /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-149018899/testdata/0/0/9/37: no such file or directory
error walking dir testdata/0/0/9/37: Lstat: lstat /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-049004756/testdata/0/0/9/37: no such file or directory
error walking dir 0/9: Open: open /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-184104134/testdata/0/0/9: no such file or directory
 ID          User        Host        Created
----------------------------------------------------------------------
*dc3fff31    lloeki      Galileo.local  2017-07-06 17:28:03
foo: ModTime does not match (2017-07-06 17:28:04.513803084 +0200 CEST != 2017-07-06 17:28:04.513804 +0200 CEST)
%foo
foo/bar: ModTime does not match (2017-07-06 17:28:08.71747536 +0200 CEST != 2017-07-06 17:28:08.717476 +0200 CEST)
%foo/bar
foo/bar/testfile0: ModTime does not match (2017-07-06 17:28:05.227860691 +0200 CEST != 2017-07-06 17:28:05.227861 +0200 CEST)
%foo/bar/testfile0
foo/bar/testfile1: ModTime does not match (2017-07-06 17:28:05.71512518 +0200 CEST != 2017-07-06 17:28:05.715126 +0200 CEST)
%foo/bar/testfile1
foo/bar/testfile2: ModTime does not match (2017-07-06 17:28:05.807257682 +0200 CEST != 2017-07-06 17:28:05.807258 +0200 CEST)
%foo/bar/testfile2
foo/bar/testfile3: ModTime does not match (2017-07-06 17:28:06.307714912 +0200 CEST != 2017-07-06 17:28:06.307715 +0200 CEST)
%foo/bar/testfile3
foo/bar/testfile4: ModTime does not match (2017-07-06 17:28:07.078611832 +0200 CEST != 2017-07-06 17:28:07.078612 +0200 CEST)
%foo/bar/testfile4
foo/bar/testfile5: ModTime does not match (2017-07-06 17:28:07.604224415 +0200 CEST != 2017-07-06 17:28:07.604225 +0200 CEST)
%foo/bar/testfile5
foo/bar/testfile6: ModTime does not match (2017-07-06 17:28:07.971472573 +0200 CEST != 2017-07-06 17:28:07.971473 +0200 CEST)
%foo/bar/testfile6
foo/bar/testfile7: ModTime does not match (2017-07-06 17:28:08.023946232 +0200 CEST != 2017-07-06 17:28:08.023947 +0200 CEST)
%foo/bar/testfile7
foo/bar/testfile8: ModTime does not match (2017-07-06 17:28:08.708402862 +0200 CEST != 2017-07-06 17:28:08.708403 +0200 CEST)
%foo/bar/testfile8
foo/bar/testfile9: ModTime does not match (2017-07-06 17:28:09.223617926 +0200 CEST != 2017-07-06 17:28:09.223618 +0200 CEST)
%foo/bar/testfile9
integration_test.go:893: directories are not equal

--- FAIL: TestRestore (6.17s)
	testing.go:29: using low-security KDF parameters for test
	testing.go:29: using low-security KDF parameters for test
	lock.go:77: setting lock timeout to 0s
	integration_test.go:51: repository initialized at /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-030682587/repo
	integration_test.go:55: backing up [/var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-030682587/testdata]
index d5697fe0 has old format
There were 4 errors
index d5697fe0 has old format
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir: operation not permitted
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar: operation not permitted
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar2: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/bar2: operation not permitted
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/baz: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-265310113/restore/testdir/baz: operation not permitted
There were 4 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
ignoring error for /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: Lchown: lchown /var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/restic-test-062584457/restore/random: operation not permitted
There were 1 errors
FAIL
FATAL: command "test" failed: [/var/folders/8n/00j6g_xx08qb21g22s_jj7sh0000gn/T/gb442135313/cmds/restic/testmain/_test/testmain]: exit status 1
```

## Steps to reproduce the behavior

run tests on 10.13 with APFS

## Do you have any idea what may have caused this?

APFS is known to have increased timestamp precision down to nanosecond.

**Comment by fd0** [comment]
Oh, interesting, thanks for raising this issue! I must say, I don't use OS X and I don't have access to a machine with it, but it sounds like we can resolve this without me having direct access. I'll have a look.

**Comment by middelink** [comment]
Why is the travis osx run not showing this? Not using APFS perhaps?

**Comment by fd0** [comment]
Yeah, I suppose they aren't using APFS yet. As far as I know it's only in beta right now.

**Comment by fd0** [comment]
It feels to me that somewhere the timestamp is truncated, maybe in the JSON encoding?

**Comment by fd0** [comment]
Hm, doesn't look that way. Strange.

**Comment by fd0** [comment]
@lloeki Could you please run the integration test again with debug?
```
$ DEBUG_LOG=/tmp/log gb test -tags debug -v -run 'TestRestore$' cmds/restic
```

(absolute path in `$DEBUG_LOG` is important)

**Comment by lloeki** [comment]
@fd0 sure, output:

```
> DEBUG_LOG=/tmp/log gb test -tags debug -v -run 'TestRestore$' cmds/restic
restic/repository
restic/checker
github.com/pkg/sftp
restic/backend/sftp
restic/backend/location
# cmds/restic
debug log file /tmp/log
debug enabled
=== RUN   TestRestore
foo: ModTime does not match (2017-07-08 12:31:40.164913568 +0200 CEST != 2017-07-08 12:31:40.164914 +0200 CEST)
%foo
foo/bar: ModTime does not match (2017-07-08 12:31:44.814164684 +0200 CEST != 2017-07-08 12:31:44.814165 +0200 CEST)
%foo/bar
foo/bar/testfile0: ModTime does not match (2017-07-08 12:31:40.913921105 +0200 CEST != 2017-07-08 12:31:40.913922 +0200 CEST)
%foo/bar/testfile0
foo/bar/testfile1: ModTime does not match (2017-07-08 12:31:41.471042803 +0200 CEST != 2017-07-08 12:31:41.471043 +0200 CEST)
%foo/bar/testfile1
foo/bar/testfile2: ModTime does not match (2017-07-08 12:31:41.574751352 +0200 CEST != 2017-07-08 12:31:41.574752 +0200 CEST)
%foo/bar/testfile2
foo/bar/testfile3: ModTime does not match (2017-07-08 12:31:42.152517535 +0200 CEST != 2017-07-08 12:31:42.152518 +0200 CEST)
%foo/bar/testfile3
foo/bar/testfile4: ModTime does not match (2017-07-08 12:31:43.014842612 +0200 CEST != 2017-07-08 12:31:43.014843 +0200 CEST)
%foo/bar/testfile4
foo/bar/testfile5: ModTime does not match (2017-07-08 12:31:43.602873562 +0200 CEST != 2017-07-08 12:31:43.602874 +0200 CEST)
%foo/bar/testfile5
foo/bar/testfile6: ModTime does not match (2017-07-08 12:31:44.001077844 +0200 CEST != 2017-07-08 12:31:44.001078 +0200 CEST)
%foo/bar/testfile6
foo/bar/testfile7: ModTime does not match (2017-07-08 12:31:44.058968731 +0200 CEST != 2017-07-08 12:31:44.058969 +0200 CEST)
%foo/bar/testfile7
foo/bar/testfile8: ModTime does not match (2017-07-08 12:31:44.798702627 +0200 CEST != 2017-07-08 12:31:44.798703 +0200 CEST)
%foo/bar/testfile8
foo/bar/testfile9: ModTime does not match (2017-07-08 12:31:45.355062584 +0200 CEST != 2017-07-08 12:31:45.355063 +0200 CEST)
%foo/bar/testfile9
integration_test.go:893: directories are not equal

--- FAIL: TestRestore (7.62s)
	testing.go:29: using low-security KDF parameters for test
	testing.go:29: using low-security KDF parameters for test
	lock.go:77: setting lock timeout to 0s
	integration_test.go:51: repository initialized at /var/folders/rm/7k1v_5pj6sxgm2mq0k9fpj6h0000gn/T/restic-test-184666079/repo
	integration_test.go:55: backing up [/var/folders/rm/7k1v_5pj6sxgm2mq0k9fpj6h0000gn/T/restic-test-184666079/testdata]
FAIL
FATAL: command "test" failed: [/var/folders/rm/7k1v_5pj6sxgm2mq0k9fpj6h0000gn/T/gb422824338/cmds/restic/testmain/_test/testmain-debug -test.v=true -test.run=TestRestore$]: exit status 1
```

Contents of `/tmp/log` [attached](https://github.com/restic/restic/files/1132899/log.txt)

**Comment by rawtaz** [comment]
Now that macOS 10.13 ("High Sierra") is released, perhaps the tests should be run again, so it's done on a non-beta filesystem?

**Comment by wojas** [comment]
PR #1389 has a workaround to make the tests pass.

This issue should probably be kept open until Go's syscall is fixed to use the new darwin syscall that is required for restoring with nanosecond precision (see #1389 for some more details).

Even when Go's syscall is updated, builds made with current versions of Go will be limited to microsecond accuracy in restored timestamps, so this workaround will need to be kept around for a while.

**Comment by fd0** [comment]
The next step should be to raise an issue in the Go repository and propose a fix, usually it is merged fast. @wojas since I don't have a Darwin system anywhere, would you like to do that? I can assist and also handle adding the code to the Go repo (with Gerrit etc) if you like.

**Comment by wojas** [comment]
@fd0 I plan to create a fix this week. It's a bit more involved than just adding a new syscall, as darwin implements it through the `fsetattrlist` syscall instead of a separate `utimensat` as *BSD does. 

I will let you know if I need help, thanks!

**Comment by wojas** [comment]
@fd0 And maybe we should reopen this issue? It was automatically closed because of the PR description claiming it's a fix.

**Comment by fd0** [comment]
Good point, reopening!

**Comment by wojas** [comment]
I was about to fix this and found it was fixed only four days ago:

https://github.com/golang/go/issues/22528
https://github.com/golang/go/commit/eb2b0ed5b59cfb340de0118235bfda60e0de4a8d

How about keeping this issue open until it lands in a Go release, and then add a Go version check in the test timestamp comparison code?

**Comment by fd0** [comment]
Agreed, thanks!

**Comment by MichaelEischer** [comment]
Restic requires at least go 1.11 by now. This underlying problem of this issue and #1307 are fixed since go 1.10. Thus it should be possible to remove the macOS and OpenBSD special case again from the restorer testcase.
